juseppe-cli/src/main/java/ru/lanwen/jenkins/juseppe/files/WatchFiles.java

```diff
@@ -18,60 +26,121 @@
-        return new WatchFiles().configureFor(props);
-    }
-
-    @Override
-    public void run() {
-        LOG.info("Start to watch for changes: {}", path);
-        try {
-            // get the first event before looping
-            WatchKey key = watcher.take();
-            while (key != null) {
-
-                if (key.pollEvents().stream().anyMatch(hasExt(".hpi").or(hasExt(".jpi")))) {
-                    LOG.trace("HPI (JPI) list modify found!");
-                    UpdateSiteGen.updateSite(props).withDefaults().toSave().saveAll();
-                }
-
-                key.reset();
-                key = watcher.take();
+   private static final Logger LOG = LoggerFactory.getLogger(WatchFiles.class);
+
+   private WatchService watcher;
+   private Map<WatchKey, Path> keys;
+   private Path path;
+   private Props props;
+
+   private WatchFiles() {
+      setDaemon(true);
+   }
+
+   public WatchFiles configureFor(Props props) throws IOException {
+      this.props = props;
+      this.path = Paths.get(props.getPluginsDir());
+      this.keys = new HashMap<>();
+      setName(format("file-watcher-%s", path.getFileName()));
+
+      watcher = this.path.getFileSystem().newWatchService();
+      walkAndRegisterDirectories(path);
+
+      return this;
+   }
+
+   public static WatchFiles watchFor(Props props) throws IOException {
+      return new WatchFiles().configureFor(props);
+   }
+
+   /**
+    * Register the given directory with the WatchService; This function will be called by FileVisitor
+    */
+   private void registerDirectory(Path dir) throws IOException {
+      WatchKey key = dir.register(watcher, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY);
+      keys.put(key, dir);
+   }
+
+   /**
+    * Register the given directory, and all its sub-directories, with the WatchService.
+    */
+   private void walkAndRegisterDirectories(final Path start) throws IOException {
+      // register directory and sub-directories
+      Files.walkFileTree(start, new SimpleFileVisitor<Path>() {
+         @Override
+         public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+            registerDirectory(dir);
+            return FileVisitResult.CONTINUE;
+         }
+      });
+   }
+
+   @Override
+   public void run() {
+      LOG.info("Start to watch for changes: {}", path);
+      while (true) {
+         // wait for key to be signalled
+         WatchKey key;
+         try {
+            key = watcher.take();
+         } catch (InterruptedException x) {
+            LOG.debug(String.format("%s: Cancelled watch service", getClass()));
+            break;
+         }
+
+         Path dir = keys.get(key);
+
+         if (dir == null) {
+            LOG.error(String.format("%s: WatchKey not recognized!!", getClass()));
+            continue;
+         }
+
+         for (WatchEvent<?> event : key.pollEvents()) {
+            @SuppressWarnings("rawtypes")
```

Comment:
put this on the method - its not very helpful to see it in the middle of the method